### PR TITLE
enhancement(sources): make scrape interval configurable

### LINKED ISSUE #4997: Make the scrape interval of internal metrics configurable
The `internal_metrics` has a hardcoded scrape interval of every 2 seconds. Would like to make that configurable so it can be controlled in a push based metrics model.

### PR DESCRIPTION
**Pull request by drunkirishcoder**
closes #4997

### CODE REVIEW

**Comment by jamtur01** [comment]
Nice. +1.

File: src/sources/internal_metrics.rs [hunk 1]
```diff
@@ -13,16 +13,30 @@ use futures::{
...
+pub fn default_scrape_interval_secs() -> u64 {
+    2
+}
 
 inventory::submit! {
     SourceDescription::new::<InternalMetricsConfig>("internal_metrics")
 }
 
-impl_generate_config_from_default!(InternalMetricsConfig);
+impl GenerateConfig for InternalMetricsConfig {
```
Comment (jszwedko) [inline src/sources/internal_metrics.rs:32]: I think we could implement `Default` here instead. That would have the advantage of also being able to use `InternalMetricsConfig::default()` in tests and other places.

File: src/sources/internal_metrics.rs [hunk 2]
```diff
@@ -13,16 +13,30 @@ use futures::{
...
+use tokio::select;
 
 #[derive(Deserialize, Serialize, Debug, Clone, Default)]
-pub struct InternalMetricsConfig {}
+pub struct InternalMetricsConfig {
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
```
Comment (jszwedko) [inline src/sources/internal_metrics.rs:24]: I think this can be a `const fn`.

**Review by jszwedko — APPROVED** [review]
Thanks @drunkirishcoder ! Two minor comments, but otherwise this looks good.

**Comment by jszwedko** [comment]
Looks good!

I added a ` #[serde(deny_unknown_fields)]` when I realized a typo in my config was causing this setting to not be applied when I was testing it out.

I'll merge once CI is green.

**Comment by jszwedko** [comment]
Test failures seem unrelated:

* Windows OOM issue we've been seeing recently

OSX:

```
 test config::watcher::tests::file_update ... Nov 11 22:12:12.133  INFO vector::config::watcher: Creating configuration file watcher.
Nov 11 22:12:12.136  INFO vector::config::watcher: Watching configuration files.
thread 'main' panicked at 'Test timed out', src/config/watcher.rs:150:13
```

**Pull request merged**